File: web-app/irp-web-app/src/GestureHeuristics.js
//BabylonJs for Vector maths
import {Vector3} from "@babylonjs/core";
import LandmarksHelper from './LandmarksHelper';

export default class GestureHeuristics {
	
	constructor(){
        //variables for previous middle tip position
        this.prevMiddleTipPosLeft = null;
        this.prevMiddleTipPosRight = null;
        //variables for last prediction time
        this.lastPredictionLeft = null;
        this.lastPredictionRight = null;
        //bind predict methods to this class
        this.predictLeft = this.predictLeft.bind(this);
        this.predictRight = this.predictRight.bind(this);

        //landmarks helper
        this.landmarksHelper = new LandmarksHelper();
	}

    //
    //Heuristics methods
    isThumbsUp(fingerStates,thumbTipY,knuckleY){
        if((knuckleY>0.70) && (thumbTipY>0.55) && !fingerStates[1] && !fingerStates[2] && !fingerStates[3] && !fingerStates[4])
            return true;
        
        else 
            return false
    }

    isSwear(fingerStates){
        if(!fingerStates[1] && fingerStates[2] && !fingerStates[3] && !fingerStates[4])
            return true;   
        else 
            return false
    }

    isRaiseHand(fingerStates,knuckleX,palmY,hand="Right",velocity){
        if((palmY>0.85) && 
            ((knuckleX>0.75 && hand ==="Right")||(knuckleX<-0.75 && hand ==="Left")) 
            && fingerStates[1] && fingerStates[2] && fingerStates[3] && fingerStates[4]
            && velocity < 0.01){
                return true;
            }
               
        else 
            return false
    }


    isWave(fingerStates,handVelocity,knuckleX,palmY,hand="Right"){
        //fingers straight
        //palm forward
        //velocity greater than x?     
        let v = handVelocity*1000;
        if((handVelocity>0.1) &&
            (palmY >0.5) &&
            fingerStates[1] && fingerStates[2] && fingerStates[3] && fingerStates[4] &&
            ((knuckleX>0.4 && hand ==="Right")||(knuckleX<-0.4 && hand ==="Left"))){
            return true;
        }          
        else 
            return false
    }

    isOK(fingerStates,distIT,knuckleX,palmY,hand="Right"){
        if((palmY>0.85) && 
            ((knuckleX>0.50 && hand ==="Right")||(knuckleX<-0.50 && hand ==="Left"))
            && distIT<=0.06 && !fingerStates[1] && fingerStates[2] && fingerStates[3] && fingerStates[4])
            return true;
        else 
            return false
    }

    //Predict gesture given landmarks
    predict(landmarks,hand,lastPrediction,middleTip,prevMiddleTipPos){

        //return 0,1,3,4,5,
        const fingerStatesCollinear = this.landmarksHelper.getFingerStatesCollinear(landmarks);
        
        const knuckleDir = this.landmarksHelper.getDirectionVector(landmarks[5],landmarks[17])

        const thumbTipY = this.landmarksHelper.getDirectionVector(landmarks[4],landmarks[3]).y;
        const knuckleX = knuckleDir.x;
        const knuckleY = knuckleDir.y;
        const palmY = this.landmarksHelper.getDirectionVector(landmarks[9],landmarks[0]).y;


        const indexTip = new Vector3(landmarks[8].x,landmarks[8].y,landmarks[8].z);
        const thumbTip = new Vector3(landmarks[4].x,landmarks[4].y,landmarks[4].z);
        
        const distIT = Vector3.Distance(indexTip,thumbTip);
        
        let handVelocity = 0; 

        if(prevMiddleTipPos && lastPrediction){
            let td = Date.now() - lastPrediction;
            handVelocity = this.landmarksHelper.getXVelocity(middleTip,prevMiddleTipPos,td);
        }

        if(this.isThumbsUp(fingerStatesCollinear,thumbTipY,knuckleY)){
            return 0;         
        }
        else if(this.isSwear(fingerStatesCollinear)){
            return 5;         
        }
        else if(this.isWave(fingerStatesCollinear,handVelocity,knuckleX,palmY,hand)){
            return 4;        
        }
        else if(this.isRaiseHand(fingerStatesCollinear,knuckleX,palmY,hand,handVelocity)){
            return 1;                                      
        }
        else if(this.isOK(fingerStatesCollinear,distIT,knuckleX,palmY,hand)){
            return 2;         
        }
        else {
            return 3;        
        }       
    }

    predictLeft(landmarks,hand){
        //set up temporal left hand variables
        const lastPrediction = this.lastPredictionLeft;
        const prevMiddleTipPos = this.prevMiddleTipPosLeft;
        const middleTip = new Vector3(landmarks[12].x,landmarks[12].y,landmarks[12].z);        
        //predict
        const prediction = this.predict(landmarks,hand,lastPrediction,middleTip,prevMiddleTipPos);
        //update temporal right hand variables
        this.prevMiddleTipPosLeft = middleTip;
        this.lastPredictionLeft = Date.now();
        return prediction;
    }

    predictRight(landmarks,hand){
        //set up temporal right hand variables
        const lastPrediction = this.lastPredictionRight;
        const prevMiddleTipPos = this.prevMiddleTipPosRight;
        const middleTip = new Vector3(landmarks[12].x,landmarks[12].y,landmarks[12].z);
        //predict
        const prediction = this.predict(landmarks,hand,lastPrediction,middleTip,prevMiddleTipPos);
        //update temporal right hand variables
        this.prevMiddleTipPosRight = middleTip;
        this.lastPredictionRight = Date.now();
        
        return prediction;
    }
}

File: python-scripts/gestureHeuristics.py
import numpy as np
from gestureFeatures import get_direction, get_finger_states, get_x_velocity

'''Class for calculating and predicting gesture heuristics'''

class GestureHeuristics:

    prev_pos = np.zeros(3)

    '''Heuristics functions'''

    def isThumbsUp(self,fingerStates,thumbTipY,knuckleY):
        if knuckleY>0.70 and thumbTipY>0.55 and not fingerStates[0] and not fingerStates[1] and not fingerStates[2] and not fingerStates[3]:
            return True    
        else:
            return False

    def isRaiseHand(self,fingerStates,knuckleX,palmY,hand,velocity=0):
        if (palmY>0.85) and ((knuckleX>0.75 and hand == "Right") or (knuckleX<-0.75 and hand == "Left")) and fingerStates[0] and fingerStates[1] and fingerStates[2] and fingerStates[3]:
            return True    
        else:
            return False

    def isOK(self,fingerStates,distIT,knuckleX,palmY,hand):
        if ((palmY>0.85) and 
            ((knuckleX>0.50 and hand == "Right") or (knuckleX<-0.50 and hand == "Left")) 
            and distIT <= 0.06 
            and not fingerStates[0] and fingerStates[1] and fingerStates[2] and fingerStates[3]):
            return True    
        else:
            return False

    def isWave(self,fingerStates,handVelocity,knuckleX,palmY,hand):
        if ((handVelocity>0.1) and
            (palmY >0.5) and
            ((knuckleX>0.4 and hand == 1) or (knuckleX<-0.4 and hand == 0)) and
            fingerStates[0] and fingerStates[1] and fingerStates[2] and fingerStates[3]):
            return True    
        else:
            return False

    #Used to clear previous position each time a new sequence begins
    def clear_prev_pos(self):
        self.prev_pos = np.zeros(3)

    '''Predict gesture given landmarks and handeness'''

    def predict(self,landmarks,hand,static_only = False):
        #finger states array
        fingerStates = get_finger_states(landmarks)

        #directions
        knuckleDir = get_direction(landmarks[5],landmarks[17])
        knuckleX = knuckleDir[0]
        knuckleY = knuckleDir[1]
        palmY = get_direction(landmarks[9],landmarks[0])[1]
        thumbTipY = get_direction(landmarks[4],landmarks[3])[1]

        #distance between index and thumb tip
        indexTip = np.array([landmarks[8]['x'],landmarks[8]['y'],landmarks[8]['z']])
        thumbTip = np.array([landmarks[4]['x'],landmarks[4]['y'],landmarks[4]['z']])                
        distIT = np.linalg.norm(indexTip-thumbTip)

        #x velocity middle finger tip
        middleTip = np.array([landmarks[12]['x'],landmarks[12]['y'],landmarks[12]['z']])
        
        #calculate velocity assuming 30fs
        mx_velocity = get_x_velocity(middleTip,self.prev_pos,33.33)
        self.prev_pos = middleTip

        #Test each heuristic
        if self.isThumbsUp(fingerStates,thumbTipY,knuckleY):
            return "thumbs-up"
        elif self.isRaiseHand(fingerStates,knuckleX,palmY,hand):
            return "raise-hand"
        elif self.isOK(fingerStates,distIT,knuckleX,palmY,hand):
            return "ok"
        elif self.isWave(fingerStates,mx_velocity,knuckleX,palmY,hand) and not static_only:
            return "wave"
        else:
            return "non-gesture"